Protect card: Add the Last scan time (with tooltip) based on plugin state.

elliottprogrammer opened the pull request:
This PR is the first PR in a series that will add/show various pieces of information on the Protect product card in My Jetpack. In this PR we will show:

1. if the Protect plugin is Not active, we will show the number of installed plugins and themes on the site. For Example: "12 Plugins & 3 Themes ⓘ" (with an info-icon popup).
2. If the Protect plugin is active, we will show the last scan time, For example: "Last scan: 17 hours ago".

See Design P2: p1HpG7-rFA-p2 (Figma link is within the post)

### Screenshots:

**Protect plugin _Not active_:**

![Screen Shot 2024-06-27 at 12 06 26](https://github.com/Automattic/jetpack/assets/11078128/4aed7300-2c73-4964-a1a4-918bdb29ef80)

![Screen Shot 2024-06-27 at 12 07 29](https://github.com/Automattic/jetpack/assets/11078128/f85b3a0e-ad6c-4d68-bd5e-15e4eebb9876)


**Protect plugin _Active_:**

![Screen Shot 2024-06-27 at 12 04 49](https://github.com/Automattic/jetpack/assets/11078128/ab81007f-faaf-40a3-8135-ce44ce301ea4)


<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

**Note:** The "Scan", "Auto-Firewall", and "Logins-Blocked" sections are currently empty and not functional. Those will be added to in upcoming PR's.

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Adds a 'themes' property to the `window.myJetpackInitialState` object.
* Adds a `protectStatus` property to the `window.myJetpackInitialState` object.
* Imports/uses the `protect-status` package.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

Project Thread: pbNhbs-aP6-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
 No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

- Run this PR locally (Docker dev environment) or from a Jurassic Ninja site using the Beta Tester plugin running this Jetpack branch.
- First make sure the Protect plugin is installed, but **Not** active.
- Go to My Jetpack and view the Protect product card.
- Verify you see the number of installed plugins and themes of the site on the Protect card (upper right corner).
- Verify you see an info icon (tooltip) and verify the info popup/tooltip works correctly and it matches the Figma design (linked in the Design post: p1HpG7-rFA-p2)
- Activate the Protect plugin. (Also make sure Jetpack is connected)
- Go back to My Jetpack and view the Protect card again.
- Verify you now see the "Last scan time:" with a proper value (in the upper right corner of the card)
    - On a brand new site, you may need to wait a minute for the initial first scan to complete, and then refresh the page to see the "Last scan time:" with a proper value. (although typically the first scan is very quick (just a couple seconds), in my experience).
- You can verify and compare the design with the Figma design (linked in the Design post: p1HpG7-rFA-p2)

elliottprogrammer committed:
Protect card: Add the Last scan time (with tooltip) based on plugin state.

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/index.tsx:
```diff
@@ -0,0 +1,23 @@
+
+const ProtectCard: FC< { admin: boolean } > = ( { admin } ) => {
+	// Override the primary action button to read "Protect your site" instead
+	// of the default text, "Lern more".
```
Minor typo here
```suggestion
	// of the default text, "Learn more".
```

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/protect-value-section.tsx:
```diff
@@ -0,0 +1,113 @@
+const ProtectValueSection: React.FunctionComponent = () => {
+	const slug = 'protect';
+	const { detail } = useProduct( slug );
+	const { status } = detail;
```
Might want to add something like this, just in case something weird happens and the detail is null instead of an object
```suggestion
	const { status = PRODUCT_STATUSES.INACTIVE } = detail || {};
```

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/protect-value-section.tsx:
```diff
@@ -0,0 +1,113 @@
+	const { detail } = useProduct( slug );
+	const { status } = detail;
+	const isProtectActive =
+		status === PRODUCT_STATUSES.ACTIVE || status === PRODUCT_STATUSES.CAN_UPGRADE;
```
We might want to group statuses into "Active", "Inactive", and "Broken" statuses like we do on the backend for checks like these 🤔

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/protect-value-section.tsx:
```diff
@@ -0,0 +1,113 @@
+
+export default ProtectValueSection;
+
+const WithProtectValueSection: React.FunctionComponent = () => {
```
Might want to import FC directly. Also it may not be necessary in this case since TypeScript should detect the correct type here automatically 🤔. Personally I like using FC when there are props because it looks nice in the code 😅 but there was a P2 written recently that suggests FC by itself may not be useful: pdWQjU-Np-p2

elliottprogrammer replied:
Yeah, I'll use FC where there are props in use. 👍

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/protect-value-section.tsx:
```diff
@@ -0,0 +1,113 @@
+	lastScanText: string;
+} > = ( { isProtectActive, lastScanText } ) => {
+	const isMobileViewport: boolean = useViewportMatch( 'medium', '<' );
+	const [ isVisiblePopover_PluginsThemes, setIsVisiblePopover_PluginsThemes ] = useState( false );
```
Mixing snake case and camel case is not a standard practice that I've seen personally 🤔 

Since these variables are restricted to this file, I think it would be okay to just name them `isPopoverVisible` and `setIsPopoverVisible`

CodeyGuyDylan replied:
Same with `togglePopover_PluginsThemes`

elliottprogrammer replied:
Yeah, hmm.. I only did this because there is going to be three(3) different info popovers on the card.  1 for PluginsThemes, 1 for ScansThreats, and 1 for AutoFirewall, each of which will need their own 'visible' state.

Right now for this PR, while there is only 1 popover, I'll change this over to `isPopoverVisible` and `setIsPopoverVisible` (and `togglePopover()`) for now.
But then in following PR's when I add the additional popovers I'll need to come up with a better way to keep track of the state of all of the popovers.  Ideally, only one popover should be able to be opened at a time, meaning that when the user opens a popover on the card, any other popovers that may be previously opened, should close first before opening the currently selected popover, if that makes sense.. Anyway I'll deal with that in upcoming PR(s).

CodeyGuyDylan replied:
Oh in that case it's okay to specify which popover it is, but we should probably stick to just camel case when naming it 😄

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/protect-value-section.tsx:
```diff
@@ -0,0 +1,113 @@
+	const [ isVisiblePopover_PluginsThemes, setIsVisiblePopover_PluginsThemes ] = useState( false );
+
+	const togglePopover_PluginsThemes = function () {
+		setIsVisiblePopover_PluginsThemes( state => ! state );
```
Nitpick, I like naming the state something more descriptive. I know the function makes it obvious what its doing but I think `previouslyVisible => ! previouslyVisible` or even just `prevState => ! prevState` would be a bit more accurate

CodeyGuyDylan replied:
We'll also want to add a tracks event here to track the popover opens and closes

elliottprogrammer replied:
Yes fair enough, sounds good! 🙂

> We'll also want to add a tracks event here to track the popover opens and closes

Yeah, my thought was to wait until I add the other popovers, to add the tracks events.  There may end up being a way to combine the tracks events for all 3 popovers into one function or something. When all 3 popovers are in place, that will be easier to determine, as well as maybe easier to determine the tracks event names and so forth.

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/protect-value-section.tsx:
```diff
@@ -0,0 +1,113 @@
+									</p>
+									<p className="value-section__tooltip-content">
+										{ __(
+											'Your site has 14 plugins and 3 themes lacking security measures. Improve your site’s safety by adding protection at no cost.',
```
I see the 14 and 3 are hardcoded here. Are these going to be changed later? I don't really see a need to postpone it when we are dynamically showing them on the card itself

elliottprogrammer replied:
Yes, I've refactored some of this PR to show the dynamic data within the tooltips. 👍

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/style.scss:
```diff
@@ -0,0 +1,51 @@
+        margin-top: var(--spacing-base);
+        column-gap: 1px;
+        position: absolute;
+        right: calc(var(--spacing-base)* 3);
```
Nitpick, weird looking spacing here
```suggestion
        right: calc(var(--spacing-base)*3);
```

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/style.scss:
```diff
@@ -0,0 +1,51 @@
+        align-items: center;
+        background: transparent;
+        border: none;
+        color: #888888;
```
It seems like this is a different grey than the one being used on the text next to it. Is there any reason we can't use the same grey for both?

elliottprogrammer replied:
Good point!  Hmm, not sure where I got that color from. 🤔
Yes, we should make them the same. 👍

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/style.scss:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    &__tooltip-heading {
+        font-size: 20px;
```
I think you can use `--font-title-small` instead

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/style.scss:
```diff
@@ -0,0 +1,51 @@
+    &__tooltip-content {
+        font-size: var(--font-body);
+        line-height: var(--font-title-small);
+        color: #444;
```
I think we should be using an existing Jetpack grey for the text. Maybe the same that's used for the text on the cards? Looks like that is `--jp-gray-70`

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/style.scss:
```diff
@@ -0,0 +1,51 @@
+        font-size: var(--font-body);
+        line-height: var(--font-title-small);
+        color: #444;
+        font-weight: 400;
```
I don't think the font we use has a specific weight for 400 🤔  When I toggle `font-weight: normal` off and on I don't see a difference

CodeyGuyDylan commented on projects/packages/my-jetpack/src/class-initializer.php:
```diff
@@ -255,6 +259,7 @@ public static function enqueue_scripts() {
 					'isAgencyAccount' => Jetpack_Manage::is_agency_account(),
 				),
 				'latestBoostSpeedScores' => $latest_score,
+				'protectStatus'          => $protect_status,
```
The name protectStatus might be misleading since we have so much code related to statuses on the frontend and backend 🤔 Maybe it could be called `protectDetails` or something similar?

elliottprogrammer replied:
Yeah, agreed..
How about `scanData`?, after all, it is the data retrieved from the `/scan` endpoint, and the data is all related to the scan of a site.

CodeyGuyDylan reviewed (changes requested):
With the Popover component, it doesn't close when I click outside of it, and it **does** close when I click inside of it. The useOutsideAlerter hook is good to handle this

https://github.com/Automattic/jetpack/assets/65001528/921ff362-11b0-426c-906e-e113468156d5

Additionally, the testing instructions say to make sure it works correctly. I am not sure if this encompasses the content of the Popover but the numbers inside are incorrect

![image](https://github.com/Automattic/jetpack/assets/65001528/d6e18b6b-09f4-4675-9c18-71fa4d78dabb)

We might want to hide the "Last scanned" text entirely until we have a scan time to show 🤔 

![image](https://github.com/Automattic/jetpack/assets/65001528/b3192c1f-fa9b-4c24-8f8f-f579d666b3aa)

I'm seeing a responsiveness issue on medium-sized screen:
![image](https://github.com/Automattic/jetpack/assets/65001528/9bffedd5-f323-4fb3-915a-ef92187b064d)

We might want to increase the minimum width of these cards as we add more information to them. It's a grid so it should scale nicely if we do that

elliottprogrammer committed:
Fix "Lern more" typo in comment, per feedback.

elliottprogrammer committed:
Add defaults to destructured var, per feedback.

elliottprogrammer committed:
Use React "FC" type, per feedback.

elliottprogrammer committed:
Fix snake case naming & state naming, per feedback.

elliottprogrammer committed:
CSS style fixes/updates, per feedback.

elliottprogrammer committed:
Show dynamic data in tooltips, per feedback; And cleanup/refactor.

elliottprogrammer committed:
Close tooltip/popover when outside focus, per feedback.

elliottprogrammer committed:
Fix tooltip close fires twice when tooltip focus switches to button clicked.

elliottprogrammer committed:
Fix responsiveness, per feedback.

elliottprogrammer committed:
Only show "Last scan:" if value available, per feedback.

elliottprogrammer committed:
Merge branch 'trunk' into add/mj-protect-card-last-scan-time

elliottprogrammer commented:
Thanks for the excellent review @CodeyGuyDylan, I've addressed all feedback. 🙌

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/protect-value-section.tsx:
```diff
@@ -0,0 +1,127 @@
+							<Gridicon icon="info-outline" size={ 14 } />
+						</button>
+						{ isPopoverVisible && (
+							<Popover
```
I think the Popover should have a `cursor: pointer` CSS line applied to it. It's not super clear that you can click on it otherwise since it's so small 🤔

elliottprogrammer replied:
Ahh, excellent point. Good catch, thank you.

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/protect-value-section.tsx:
```diff
@@ -0,0 +1,127 @@
+		[ setIsPopoverVisible ]
+	);
+	const hideTooltip = useCallback( () => {
+		// Don't hide the tooltip here if it's the toolTip button that was clicked (the button
```
The most nitpick-y of nitpicks 😅 
```suggestion
		// Don't hide the tooltip here if it's the tooltip button that was clicked (the button
```

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/protect-value-section.tsx:
```diff
@@ -0,0 +1,127 @@
+	const hideTooltip = useCallback( () => {
+		// Don't hide the tooltip here if it's the toolTip button that was clicked (the button
+		// becoming the document's activeElement). Instead let toggleTooltip() handle the closing.
+		if ( useTooltipRef.current && ! useTooltipRef.current.contains( document.activeElement ) ) {
```
This looks pretty similar to the `useOutsideAlerter` hook that already exists in this project, was there a reason it didn't work for this use-case?

elliottprogrammer replied:
Yes, there is a feature already built into the Popover that detects when the focus is triggered outside of the popover, and when that happens, the `onClose` function is fired.  The `onClose` is a user defined function via the `onClose` prop of the Popover.
So in other words, the feature is already built into the Popover, I just had to modify the `onClose` behavior slightly by telling it to not treat clicking the info-icon button as an outside focus event.

CodeyGuyDylan replied:
Ah ok I see, thank you for taking care of that!

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/style.scss:
```diff
@@ -0,0 +1,53 @@
+    justify-content: space-between;
+
+    &__heading {
+        font-size: 12px;
```
Sorry I didn't catch this last time, it looks like we can use `--font-body-extra-small` for this 🤔

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/use-protect-tooltip-copy.ts:
```diff
@@ -0,0 +1,69 @@
+			),
+		},
+		scanThreatsTooltip:
+			hasProtectPaidPlan && numThreats
```
I know this isn't going to be implemented until the next PR, but we'll want a case for a user that does not have the protect paid plan but does have threats. Right now if that were the case, the tooltip would say that all appears well even though they have threats

elliottprogrammer replied:
Ah yes thanks. Yeah, this will implemented in the next PR. 👍

CodeyGuyDylan reviewed (changes requested):
Thank you for all of the changes! It works much better for me. I did have some more comments but not nearly as many as last time 🤔  I think this is very close

Also I know we talked in slack but putting here for visibility that I'm not sure we should show the `plugins & themes` data in the event that the user has Protect active but no scan data is available due to lack of site connection or if the initial scan is still running

elliottprogrammer committed:
Add cursor: pointer to tooltip button, per feedback.

elliottprogrammer committed:
Use --font-body-extra-small & fix comment typo, per feedback.

elliottprogrammer committed:
Only show "Plugins & Themes" when scan data availavle, per feedback.

elliottprogrammer committed:
Fixup project versions.

elliottprogrammer committed:
Merge branch 'trunk' into add/mj-protect-card-last-scan-time

elliottprogrammer commented:
> Thank you for all of the changes! It works much better for me. I did have some more comments but not nearly as many as last time 🤔 I think this is very close
> 
> Also I know we talked in slack but putting here for visibility that I'm not sure we should show the `plugins & themes` data in the event that the user has Protect active but no scan data is available due to lack of site connection or if the initial scan is still running

Thanks Dylan! Yeah, I've addressed your additional feedback items, and yes, I removed the "Plugins & Themes" text when the plugin is active but there is no scan data available (b/c initial scan in progress or site disconnected, etc). 👍
Let me know if you have any additional feedback. 😃

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/use-protect-tooltip-copy.ts:
```diff
@@ -0,0 +1,69 @@
+			text: sprintf(
+				/* translators: %1$d is the number of plugins and %2$d is the number of themes installed on the site. */
+				__(
+					'Your site has %1$d plugins and %2$d themes lacking security measures. Improve your site’s safety by adding protection at no cost.',
```
Looks like we'll need to implement a `_n()` here to handle plurals 😄 
![image](https://github.com/Automattic/jetpack/assets/65001528/23f8cd5b-1234-450b-82c0-5e1c5065348e)
I know you're working on another PR having to do with tooltips so feel free to handle that there instead if you'd like

elliottprogrammer committed:
Merge branch 'trunk' into add/mj-protect-card-last-scan-time

elliottprogrammer committed:
Merge branch 'trunk' into add/mj-protect-card-last-scan-time

elliottprogrammer committed:
Merge branch 'trunk' into add/mj-protect-card-last-scan-time
